### ISSUE TITLE
PMTiles area download mechanism

### ISSUE THREAD

**Issue by wipfli** [maplibre/maplibre-native#2352]
**Is your feature request related to a problem? Please describe.**
I'm always frustrated when downloading tiles for offline usage is slow.

**Describe the solution you'd like**
The PMTiles storage format orders tiles that are on a Hilbert curve in space linearly in memory. As a consequence, tiles that are close together can be downloaded in only a few range requests whereas with traditional zxy endpoints each tile has to be downloaded individually.

So, area downloads of tiles for offline usage are much more efficient with PMTiles than with zxy endpoints.

It could be interesting for offline apps to create a mechanism that downloads tiles from a PMTiles archive and stores them in the existing offline cache...

**Describe alternatives you've considered**
The current offline download mechanism is a viable existing alternative.

**Additional context**
Measurements of @bdon have shown that `pmtiles extract...` can be extremely fast...

**Comment by louwers** [comment]
This is an interesting idea. Offline downloads of large areas are horribly slow currently indeed.

Though there would be some complexities involved in converting from PMTiles to MBTiles for storage?

**Comment by acalcutt** [comment]
That https://www.npmjs.com/package/mapgl-tile-renderer tool could probably be used for small area exports. it exports mbtiles but I added pmtiles source support to it . you don't really need the rendering aspect, but if you are supplying a style with rasters the effect is pretty much the same.

Does maplibre have a tool now to download images for offline use now?

**Comment by bdon** [comment]
Can this be accomplished outside of the maplibre native codebase, and then just deliver the tile archive in a format MLN understands?

**Comment by louwers** [comment]
@bdon Yes you can merge in a database that exists on the local filesystem. https://github.com/maplibre/maplibre-native/blob/cd6a0ab11fcd9b5bbfee87896a14159d0ed56256/platform/default/include/mbgl/storage/offline_database.hpp#L70

**Comment by louwers** [comment]
During today's TSC meeting @JesseCrocker mentioned that he had to implement his own offline database due to the limitations of the built-in one. This idea from @wipfli was also discussed.

I'll write some documentation about the current offline database implementation so we can get the ball rolling. Downloading an area, parsing it and writing it out to a SQLite database in the correct format should not be too complicated. I think it makes sense to re-use the current offline database it, since despite its flaws it is widely in use. https://github.com/maplibre/maplibre-native/issues/3017

**Comment by navisinghnz9** [comment]
Hi,
 
 I am working on an cross-platorm mobile app using maplibre-native-qt.
 I had been struggling to support pmtiles in the app.
 
 I was wondering what would be the best way to experiment with the code in the pull request?
 
 I can see that there is a sub-directory called maplibre-native, inside source tree of maplibre-native-qt.
 (I mean here - **maplibre-native-qt->vendor->maplibre-native** )
 
 Not sure if it will be a good practice to put folder maplibre-native with your PR changes/branch and recompile maplibre-native-qt SDK.
 Or I need to write some Qt binding to be able to use pmtiles features in the app.
  
 Thanks!

**Comment by louwers** [comment]
@navisinghnz9 PMTiles support is actually about to land with https://github.com/maplibre/maplibre-native/pull/2882

This pull request is about using PMTiles as a download mechanism for offline maps.

**Comment by navisinghnz9** [comment]
@louwers thanks you very much for referring me to the right pull request.
